src/segment-loader.js

```diff
@@ -1328,4 +1562,144 @@ export default class SegmentLoader extends videojs.EventTarget {
+      cue = this.segmentMetadataTrack_.cues[i];
+      for (j = 0; j < buffered.length; j++) {
+        cmpStart = (start > buffered.start(j)) ? start : buffered.start(j);
+        cmpEnd = (end < buffered.end(j)) ? end : buffered.end(j);
+        if (cue.startTime <= cmpEnd && cue.endTime >= cmpStart) {
+          bytes += cue.value.byteLength;
+          // we only want to count the bytes once, so break now
+          break;
+        }
+      }
+    }
+
+    return bytes;
+  }
+
+  // TODO: documentation
+  minBufferedBytes(start, end) {
+    if (!this.segmentMetadataTrack_) {
+      return 0;
+    }
+    if (end < start) {
+      return 0;
+    }
+
+    let i;
+    let j;
+    let cue;
+    let cmpStart;
+    let cmpEnd;
+    let bytes = 0;
+    let buffered = this.buffered_();
+
+    start = start === undefined ? 0 : start;
+    end = end === undefined ? Infinity : end;
+
+    for (i = 0; i < this.segmentMetadataTrack_.cues.length; i++) {
+      cue = this.segmentMetadataTrack_.cues[i];
+
+      for (j = 0; j < buffered.length; j++) {
+        cmpStart = start > buffered.start(j) ? start : buffered.start(j);
+        cmpEnd = end < buffered.end(j) ? end : buffered.end(j);
+
+        if (cue.endTime <= cmpEnd && cue.startTime >= cmpStart - 0.1) {
+          bytes += cue.value.byteLength;
+          break;
+        }
+      }
+    }
+
+    return bytes;
+  }
+
+  removeNonBufferedSegmentCues() {
+    if (!this.segmentMetadataTrack_) {
+      return 0;
+    }
+
+    let i;
+    let j;
+    let cue;
+    let cmpStart;
+    let cmpEnd;
+    let keep = true;
+    let removeCues = [];
+    let buffered = this.buffered_();
+
+    for (i = 0; i < this.segmentMetadataTrack_.cues.length; i++) {
+      cue = this.segmentMetadataTrack_.cues[i];
+      keep = false;
+
+      for (j = 0; j < buffered.length; j++) {
+        cmpStart = buffered.start(j);
+        cmpEnd = buffered.end(j);
+
+        if (cue.startTime <= cmpEnd && cue.endTime >= cmpStart) {
+          keep = true;
+          break;
+        }
+      }
+
+      if (!keep) {
+        removeCues.push(cue);
+      }
+    }
+
+    removeCues.forEach((cueToRemove) => {
+      this.segmentMetadataTrack_.removeCue(cueToRemove);
+    });
+  }
+
+  removeWholeSegments(end) {
+    if (!this.segmentMetadataTrack_) {
+      return [];
+    }
+
+    let i;
+    let j;
+    let cue;
+    let cmpEnd;
+    let buffered = this.buffered_();
+    let matchingCues = [];
+
+    end = end === undefined ? Infinity : end;
+
+    for (i = 0; i < this.segmentMetadataTrack_.cues.length; i++) {
+      cue = this.segmentMetadataTrack_.cues[i];
+
+      for (j = 0; j < buffered.length; j++) {
+        cmpEnd = end < buffered.end(j) ? end : buffered.end(j);
+
+        if (cue.endTime <= cmpEnd) {
+          matchingCues.push(cue);
+          break;
+        }
+      }
+    }
+    matchingCues.forEach((matchingCue) => {
+      this.remove(0, matchingCue.endTime);
```

Comment:
couldn't you just find the first cue that is outside the buffer and call remove from 0 to that cue instead of repeatedly removing from 0?

Reply:
Probably.